fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/ParameterEncoder.kt

```diff
@@ -0,0 +1,79 @@
+package com.github.kittinunf.fuel.core.interceptors
+
+import com.github.kittinunf.fuel.core.FoldableRequestInterceptor
+import com.github.kittinunf.fuel.core.Headers
+import com.github.kittinunf.fuel.core.Method
+import com.github.kittinunf.fuel.core.Parameters
+import com.github.kittinunf.fuel.core.RequestTransformer
+import java.net.URL
+import java.net.URLEncoder
+
+object ParameterEncoder : FoldableRequestInterceptor {
+    override fun invoke(next: RequestTransformer): RequestTransformer {
+        return inner@{ request ->
+            val contentType = request[Headers.CONTENT_TYPE].lastOrNull()
+
+            // Expect the parameters to be already encoded in the body
+            if (contentType?.startsWith("multipart/form-data") == true) {
+                return@inner next(request)
+            }
+
+            // If it can be added to the body
+            if (request.body.isEmpty() && allowParametersInBody(request.method)) {
+                if (contentType.isNullOrBlank() || contentType.startsWith("application/x-www-form-urlencoded")) {
+                    return@inner next(
+                        request
+                            .header(Headers.CONTENT_TYPE, "application/x-www-form-urlencoded")
+                            .body(encode(request.parameters))
+                            .apply { parameters = emptyList() }
+                    )
+                }
+            }
+
+            // Has to be added to the URL
+            next(
+                request
+                    .apply { url = url.withParameters(parameters) }
+                    .apply { parameters = emptyList() }
+            )
+        }
+    }
+
+    internal fun encode(parameters: Parameters) =
+        parameters
+        .flatMap { (key, values) ->
+            // Deal with arrays
+            ((values as? Iterable<*>)?.toList() ?: (values as? Array<*>)?.toList())?.let {
+                val encodedKey = "${URLEncoder.encode(key, "UTF-8")}[]"
+                it.map { value -> encodedKey to URLEncoder.encode(value.toString(), "UTF-8") }
+
+            // Deal with regular
+            } ?: values?.let {
+                listOf(URLEncoder.encode(key, "UTF-8") to URLEncoder.encode(values.toString(), "UTF-8"))
+
+            // Parameters that are null are stripped
+            } ?: emptyList()
```

Comment:
[nits]  Shall we eagerly filter the null value out first? Then, I think we can just deal with less `?.` thingy

``` Kotlin
parameters.filterNot { (key, values) -> values == null }
.flatMap {
   // as current impl
}
```

Reply:
Yes!

Reply:
Done!